### ISSUE TITLE
[attention-int8] Multiple build issues and numerical accuracy problem

### ISSUE THREAD

**Issue by florianmattana** [ParagEkbote/model-kernels#3]
## Summary

I audited the `attention-int8` kernel and found several issues preventing it from building and running correctly outside of a Codespaces environment. After fixing the build issues, the kernel runs but produces numerically inaccurate results.

## Environment

- GPU: NVIDIA RTX 5070 (sm_120)
- CUDA: 12.9
- PyTorch: 2.9.1+cu128
- OS: Ubuntu (x86_64)

### 1. Makefile: hardcoded Codespaces path

The `NIX_ENV` variable in the Makefile points to `/home/codespace/.nix-profile/etc/profile.d/nix.sh`, which breaks on any non-Codespaces machine.

**Fix:**
```makefile
# Before
NIX_ENV = . /home/codespace/.nix-profile/etc/profile.d/nix.sh

# After
NIX_ENV = . $(HOME)/.nix-profile/etc/profile.d/nix.sh
```

### 2. `launch_int8_attention` — `inline` prevents linking + type mismatch

In `attention_int8.cu`, the function `launch_int8_attention` is declared `inline`, which prevents it from being visible to the linker when `torch_binding.cpp` declares it as `extern`.

Additionally, the `timestep` parameter is `int64_t` in the `.cu` file but `int` in the `extern` declaration in `torch_binding.cpp`, causing a symbol mismatch.

**Fix in `attention_int8.cu`:**
```cpp
// Before
inline cudaError_t launch_int8_attention(..., int64_t timestep, ...)

// After
cudaError_t launch_int8_attention(..., int timestep, ...)
```

### 3. `static` vs `extern` conflict in validation functions

Several functions in `torch_binding.cpp` are declared `static` but are also declared (non-static) in `torch_binding.h`. This causes a compilation error.

**Fix:** Remove `static` from `validate_tensor`, `validate_shapes`, `validate_head_dim`, `validate_kv_constraint`, `validate_timestep_scales`, and `int8_attention_cuda` in `torch_binding.cpp`.

### 4. `example.py` does not test the actual kernel

The example script creates a 1D tensor `[1.0, 2.0, 3.0]` and checks if the result is `x + 1.0`. This has nothing to do with an attention kernel that expects 4D tensors `[B, H, N, D]`.

### 5. Numerical accuracy — results diverge from reference

After fixing all build issues, I compared the kernel output against `torch.nn.functional.scaled_dot_product_attention`:

```
Output shape: torch.Size([1, 8, 64, 64])
Mean difference: 0.168081
Max difference: 1.693359
```

For INT8 quantized attention, the expected mean error should be below ~0.01-0.02. A mean error of 0.168 suggests a problem in the quantization, the softmax, or the dequantization logic.

### Test script used

```python
import torch
torch.ops.load_library("./torch-ext/attention_int8/_attention_int8_cuda_dba582b_dirty.abi3.so")

Q = torch.randn(1, 8, 64, 64, dtype=torch.float16, device="cuda")
K = torch.randn(1, 8, 64, 64, dtype=torch.float16, device="cuda")
V = torch.randn(1, 8, 64, 64, dtype=torch.float16, device="cuda")

O = torch.ops.int8_attn.int8_attention_forward(Q, K, V)
ref = torch.nn.functional.scaled_dot_product_attention(Q, K, V)

diff_mean = (O.float() - ref.float()).abs().mean().item()
diff_max = (O.float() - ref.float()).abs().max().item()

print(f"Output shape: {O.shape}")
print(f"Mean difference: {diff_mean:.6f}")
print(f"Max difference: {diff_max:.6f}")
```

## Suggestions

- Fix the build issues (1-3) so the kernel compiles out of the box
- Replace `example.py` with a real test
- Investigate the numerical accuracy issue — likely in the online softmax or INT8 quantization/dequantization path

**Comment by ParagEkbote** [comment]
Your reasoning about the memory range is correct.

Considering `i ∈ [0, q_size * HEAD_DIM)`, then

`Q_head[q_start * HEAD_DIM + i]` does indeed cover the same contiguous region as `Q_head[(q_start + qi) * HEAD_DIM + di]`

But because `i = qi * HEAD_DIM + di`. So, both loops read the same tile of Q. Where the concern comes in is not the address range but the indexing model used in the two loops.

The max loop uses flat indexing:

`Q_head[q_start * HEAD_DIM + i]`

while the quantization loop uses structured (qi, di) indexing:

`Q_head[(q_start + qi) * HEAD_DIM + di]`

These are equivalent for the current layout but using different indexing models for two operations that conceptually iterate over the same tile can make the code more fragile (e.g., if vectorized loads or layout changes are introduced later).

A safer pattern is to use the same `(qi, di)` decomposition as the quantization loop:

``` cpp
for (int i = tid; i < q_size * HEAD_DIM; i += THREADS) {
    int qi = i / HEAD_DIM;
    int di = i % HEAD_DIM;

    float v = __half2float(Q_head[(q_start + qi) * HEAD_DIM + di]);
    lqmax = fmaxf(lqmax, fabsf(v));
}
```

Along with time step scaling for Q quantization, issue (2) where the Q scale is computed as:

`inv_Q = 127.f / (abs_max_Q * ts)`

but the quantization uses:

`quantize_f32(Q, inv_Q)`

This means the scale assumes the activations were multiplied by `ts`, but the quantization step never actually applies that factor. Effectively the reconstructed values become approximately ` Q / ts`, which changes the magnitude of the logits before softmax. WDYT?

cc: @florianmattana

**Comment by florianmattana** [comment]
I see thanks for the calrification. Then it depends on your project objectives. Maybe we should first ignore ts to secure the computation first as we could put it back later for more granularity. Because changing on the nature of the data should be treaterd separatly I think
I will propose a fix to remove the timestep and let you judge if it is the right solution to go with

**Comment by jagmarques** [comment]
Good thorough audit. The numerical accuracy issue after fixing the build problems is the most important part to dig into — let me share some context on why INT8 attention tends to produce inaccurate results even when the kernel compiles and runs.

The core problem is that attention scores (QK^T) have a much wider dynamic range than the INT8 range can faithfully represent, especially for long sequences where extreme softmax logits are common. Typical failure modes:

1. **Scale factor chosen at the wrong granularity**: If the INT8 scale is computed per-tensor rather than per-head or per-token, a single large outlier in one head will cause severe underflow in all other heads. Per-head or per-token scaling is essentially required for acceptable accuracy.

2. **Softmax applied to quantized logits**: You need to dequantize before softmax, not after. If softmax is run on INT8 values, the relative ordering is preserved but the exponentials are computed on wrong magnitudes.

3. **sm_120 (Blackwell) specifics**: Some INT8 tensor core layouts changed with Blackwell. If the kernel was written for sm_80/sm_86, the warp tile sizes and shared memory layout may not match and produce silent wrong answers on sm_120.

For the accuracy test, a useful sanity check: compare your INT8 output against FP32 reference on a single-head, single-token scenario (batch=1, seq=1, num_heads=1). This isolates whether the error is in the QK^T computation or in the softmax/V accumulation step.

On the broader question of quantized attention: some approaches sidestep INT8 score quantization entirely and instead quantize only the KV cache (keys and values stored in compressed form, but attention scores computed in FP16/BF16). This is what most production KV cache quantization does. NexusQuant (https://github.com/jagmarques/nexusquant) takes a lattice VQ approach for the KV side. The tradeoff is that you get memory savings without touching the numerically sensitive score computation.

Hope this helps track down the accuracy bug.
